### ISSUE TITLE
Pre-v0 Optimizations

### ISSUE THREAD

**Issue by z-a-f** [maida-ai/Research#33]
### 1) Don't materialize a $(T\times T)$ mask at large $T$

Even a bool mask blows up VRAM (e.g., $16k\times 16k \approx 256 \text{MB}$). Add a threshold (e.g., `max_mask_T=4096`)-above it, do **sliding block SDPA** (compute attention in overlapping blocks of size $W$, no global mask) or call **FlashAttention-2/3** with its native `window_size=(W_left,W_right)` which avoids any $(T\times T)$ tensor. ([GitHub][1])

```python
# in _compute_attention
if self.window_size >= T:
    attn_mask, is_causal = None, True
elif T <= self.max_mask_T:
    attn_mask, is_causal = self._get_window_mask(T, self.window_size, x.device), False
else:
    # option A: blockwise SDPA over sliding windows (no full mask)
    # option B: if FA2 available, call flash_attn with window_size=(W,0)
```

Also keep the **SDPA rule**: never set `attn_mask` and `is_causal=True` simultaneously; it de-optimizes and can error. (Your code already does this correctly.) ([PyTorch Docs][2])

### 2) Fuse projections into one GEMM

Replace `W_Q/W_K/W_V/B/gate` with a single `nn.Linear(d, 3d + d_s + d)` and `torch.split`. Fewer kernel launches & less DRAM traffic → typical **+10-30%** step speed.

```python
# __init__
self.W_fused = nn.Linear(d_model, 3*d_model + self.ssm_state_dim + d_model, bias=True)

# forward
fused = self.W_fused(x_norm1)  # (B,T, 3d + d_s + d)
Q, K, V, u, gate_pre = torch.split(fused, [d, d, d, self.ssm_state_dim, d], dim=-1)
g = torch.sigmoid(gate_pre)
# use u directly in SSM (drop self.B)
```

### 3) Overlap attention & SSM on **two CUDA streams**

They're independent until fusion. Run them concurrently and sync once-often hides **20-40%** of one path.

```python
if torch.cuda.is_available():
    s1, s2 = torch.cuda.Stream(), torch.cuda.Stream()
    with torch.cuda.stream(s1): y_attn = self._compute_attention(x_norm1)
    with torch.cuda.stream(s2): y_ssm, new_state = self._compute_ssm(x_norm1, ssm_state)
    e1, e2 = torch.cuda.Event(), torch.cuda.Event(); e1.record(s1); e2.record(s2)
    torch.cuda.current_stream().wait_event(e1); torch.cuda.current_stream().wait_event(e2)
else:
    ...
```

### 4) KV cache as a **ring buffer** (streaming)

Avoid `torch.cat` each step (realloc + memcpy). Preallocate `(B,h,W,d_h)` and maintain an index:

```python
# allocate once outside; keep idx on module or in attn_state
K_buf[..., idx:idx+1, :] = K_new; V_buf[..., idx:idx+1, :] = V_new
idx = (idx + 1) % W
# gather logical window by 2-slice concat if wraparound
```

### 5) Tile-scan SSM (keep your recurrence; cut the long loop)

You already precompute `u = B @ x`. Compute summaries per tile of size `C` (256/512), do a tiny prefix over tiles, then a **short in-tile loop**:

* `A_tile = alpha**C` (elementwise)
* `b_tile = sum( alpha**(C-1-i) * u_tile[i] )` (vectorized with a precomputed weight vector)
* Combine tiles via associative op $(A_2, b_2) \oplus (A_1, b_1) = (A_2 A_1, b_2 + A_2 b_1)$; this gives the **carry** for each tile.
* Enter each tile with its carry and produce per-token `C(s)` with ≤`C` steps.

Expect **2-4x** SSM-path speedup with identical math. Later, you can move to a full **parallel scan** or drop-in a **selective-scan** kernel from the Mamba family (keeps the same DP-ASSM API). ([GitHub][3])

### 6) Grouped-Query Attention (GQA)

Add `num_kv_heads` to share K/V across query groups (MQA is the `num_kv_heads=1` case). Cuts K/V bandwidth/activations with minor quality impact; widely used in production LLMs. ([arXiv][4])

```python
# after reshaping K,V: (B, n_kv, T, d_h)
repeat = self.n_heads // self.num_kv_heads
K = K.repeat_interleave(repeat, dim=1)
V = V.repeat_interleave(repeat, dim=1)
```

### 7) SSM cost knobs (great for ablations)

* **`proj_stride=K`**: compute `C(s_t)` every `K` tokens (e.g., 2-4), linearly interpolate between; always update `s_t`.
* **Layer frequency**: run SSM every `N` layers or share one SSM across a block group.
* **Low-rank B/C**: factor B and C via small inner dim `r` (`d→r→d_s`, `d_s→r→d`) to shrink GEMMs.

### 8) Use FA2's native sliding-window kernels (training & eval)

Where available, FA2's windowed path reduces memory traffic and improves throughput for local attention windows-no $(T\times T)$ mask. You can call it directly or let SDPA pick the FA backend when configured. ([GitHub][1])

### 9) Compiler + TF32

Wrap the block/model with:

```python
model = torch.compile(model, fullgraph=True, mode="reduce-overhead")
torch.set_float32_matmul_precision("high")  # enables TF32 on Ampere/Hopper
```

This fuses pointwise ops and speeds matmuls on A100/H100 without accuracy loss in most LLM setups.

### 10) Context scaling (for headline long-context demos)

If you'll benchmark 128k-1M tokens, add **RingAttention / Context Parallelism** options to scale prefill across GPUs **without approximation**. Not a single-GPU speed fix, but lets you publish strong million-token numbers and keep architecture unchanged. ([ICLR Proceedings][5])

## Stability polish (small but helpful)

* **Gate bias init** toward attention (e.g., `bias=+2.0`, so `g≈0.88` at start) for easier optimization; or toward SSM if that fits your regime.
* Consider $\alpha = \exp(-\text{softplus}(\tau))$ (contraction in (0,1)) instead of $\tanh(A)$ if you see drift in bf16. (Keep both; A/B later.)

## What you've already done right (keep it)

* **SDPA fast-path & bool mask** with correct semantics (no mask + `is_causal=True`, or mask + `is_causal=False`). ([PyTorch Docs][2])
* **Preallocated SSM outputs** and precomputed `u = B @ x` (nice).
* **Streaming KV cache** now attends over the last `W` tokens (correct).

## Minimal measurement plan (per PR)

* Report **tokens/sec** and **max VRAM** at $T\in \{512, 1024, 4096\}$, $W\in\{128, 256, 512\}$ (bf16).
* Time **attention**, **SSM**, and **other** separately (PyTorch profiler w/ CUDA time).
* For SSM changes, also time `_compute_ssm` alone.
* Keep a tiny script under `scripts/profile_dpassm.py`.

### TL;DR next steps

1. Add **mask threshold + sliding window without global mask** (or FA2 window). ([GitHub][1])
2. **Fuse projections** (QKV + u + gate).
3. **Overlap** attention & SSM via CUDA streams.
4. Implement **tile-scan SSM** (flag `ssm_impl="naive|tile_scan"`).
5. Optional: **GQA**, `proj_stride`, `torch.compile` + TF32.

[1]: https://github.com/Dao-AILab/flash-attention?utm_source=chatgpt.com "Dao-AILab/flash-attention: Fast and memory-efficient ..."
[2]: https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html?utm_source=chatgpt.com "torch.nn.functional.scaled_dot_product_attention"
[3]: https://github.com/state-spaces/mamba?utm_source=chatgpt.com "state-spaces/mamba: Mamba SSM architecture"
[4]: https://arxiv.org/abs/2305.13245?utm_source=chatgpt.com "[2305.13245] GQA: Training Generalized Multi-Query ..."
[5]: https://proceedings.iclr.cc/paper_files/paper/2024/file/1119587863e78451f080da2a768c4935-Paper-Conference.pdf?utm_source=chatgpt.com "RingAttention with Blockwise Transformers for Near-Infinite ..."
